stdlib/universe/sort_limit.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func (s *sortLimitTransformation) reconcileSchema(mh *sortTableMergeHeap, buffer *arrow.TableBuffer, mem memory.Allocator) {
+	if len(buffer.Columns) == len(mh.cols) {
+		equivalent := true
+		for i, col := range mh.cols {
+			if buffer.Columns[i] != col {
+				equivalent = false
+				break
+			}
+		}
+
+		if equivalent {
+			return
+		}
+	}
+
+	vals := make([]array.Interface, len(mh.cols))
+	for i, col := range buffer.Columns {
+		idx := execute.ColIdx(col.Label, mh.cols)
+		if idx < 0 {
+			// This column was not previously seen.
+			// Backfill the schema and add null columns
+			// in the relevant locations.
+			mh.cols = append(mh.cols, col)
+			if execute.ContainsStr(s.cols, col.Label) {
+				mh.sortCols = append(mh.sortCols, len(mh.cols)-1)
+			}
+			s.backfillColumn(mh, len(mh.cols)-1, mem)
+			vals = append(vals, buffer.Values[i])
+			continue
+		}
+		vals[idx] = buffer.Values[i]
+	}
+
+	// If a previous column existed but doesn't exist in the current buffer,
+	// we need to also backfill it.
+	for i := range vals {
+		if vals[i] == nil {
+			vals[i] = arrow.Nulls(mh.cols[i].Type, buffer.Len(), mem)
+		}
+	}
+	buffer.Columns = mh.cols
+	buffer.Values = vals
+}
+
+func (s *sortLimitTransformation) backfillColumn(mh *sortTableMergeHeap, i int, mem memory.Allocator) {
+	for _, item := range mh.items {
+		cpy := &arrow.TableBuffer{
+			GroupKey: item.cr.Key(),
+			Columns:  mh.cols,
+			Values:   make([]array.Interface, len(mh.cols)),
+		}
+		for j := range item.cr.Cols() {
+			cpy.Values[j] = table.Values(item.cr, j)
+		}
+		cpy.Values[i] = arrow.Nulls(mh.cols[i].Type, item.cr.Len(), mem)
+		item.cr = cpy
+	}
+}
+
+func (s *sortLimitTransformation) Compute(key flux.GroupKey, state interface{}, d *execute.TransportDataset, mem memory.Allocator) error {
+	// The chunks are in sorted order already and already chunked.
+	mh := state.(*sortTableMergeHeap)
+	for _, item := range mh.items {
+		chunk := table.ChunkFromReader(item.cr)
+		if err := d.Process(chunk); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *sortLimitTransformation) Close() error {
+	return nil
+}
+
+type SortLimitRule struct{}
+
+func (s SortLimitRule) Name() string {
+	return "SortLimitRule"
+}
+
+func (s SortLimitRule) Pattern() plan.Pattern {
+	return plan.Pat(LimitKind, plan.Pat(SortKind, plan.Any()))
+}
+
+func (s SortLimitRule) Rewrite(ctx context.Context, node plan.Node) (plan.Node, bool, error) {
+	if !feature.OptimizeSortLimit().Enabled(ctx) {
+		return node, false, nil
+	}
+
+	limitSpec := node.ProcedureSpec().(*LimitProcedureSpec)
+	if limitSpec.Offset != 0 {
+		return node, false, nil
+	}
+	sortNode := node.Predecessors()[0]
+	sortSpec := sortNode.ProcedureSpec().(*SortProcedureSpec)
+
+	sortLimitSpec := &SortLimitProcedureSpec{
+		SortProcedureSpec: sortSpec,
+		limit:             limitSpec.N,
+	}
+
+	n, err := plan.MergeToPhysicalNode(node, sortNode, sortLimitSpec)
+	if err != nil {
+		return nil, false, err
+	}
+	return n, true, nil
+}
```

Comment:
Could I trouble you to add some test cases for this rule?

Reply:
Yep, can do.